File: snippets/screens/swiftui/tasks_list_screen/2/tasks_list_screen.swift
import SwiftUI

struct TasksListScreen: View {
    @State private var selectedTab = 0
    @State private var showingAddTask = false
    @State private var newTaskTitle = ""

    @State private var tasks = [
        Task(
            id: "1",
            title: "Complete project proposal",
            description: "Write and submit the Q4 project proposal for the mobile app redesign",
            isCompleted: false,
            priority: .high,
            dueDate: Calendar.current.date(byAdding: .day, value: 2, to: Date()),
            category: .work
        ),
        Task(
            id: "2",
            title: "Buy groceries",
            description: "Milk, bread, eggs, fruits, and vegetables for the week",
            isCompleted: false,
            priority: .medium,
            dueDate: Calendar.current.date(byAdding: .day, value: 1, to: Date()),
            category: .personal
        ),
        Task(
            id: "3",
            title: "Review SwiftUI documentation",
            description: "Read through the latest SwiftUI documentation and new features",
            isCompleted: true,
            priority: .medium,
            dueDate: Calendar.current.date(byAdding: .day, value: -1, to: Date()),
            category: .learning
        ),
        Task(
            id: "4",
            title: "Plan weekend trip",
            description: "Research destinations, book accommodation, and plan activities",
            isCompleted: false,
            priority: .low,
            dueDate: Calendar.current.date(byAdding: .day, value: 7, to: Date()),
            category: .personal
        ),
        Task(
            id: "5",
            title: "Call dentist appointment",
            description: "Schedule routine checkup and cleaning for next month",
            isCompleted: false,
            priority: .high,
            dueDate: Calendar.current.date(byAdding: .day, value: 1, to: Date()),
            category: .health
        )
    ]

    var filteredTasks: [Task] {
        switch selectedTab {
        case 0: // All
            return tasks
        case 1: // Active
            return tasks.filter { !$0.isCompleted }
        case 2: // Completed
            return tasks.filter { $0.isCompleted }
        default:
            return tasks
        }
    }

    var activeTasksCount: Int {
        tasks.filter { !$0.isCompleted }.count
    }

    var completedTasksCount: Int {
        tasks.filter { $0.isCompleted }.count
    }

    var progressPercentage: Double {
        guard !tasks.isEmpty else { return 0 }
        return Double(completedTasksCount) / Double(tasks.count)
    }

    var body: some View {
        NavigationStack {
            VStack(spacing: 0) {
                // Progress Summary
                ProgressSummaryView(
                    completed: completedTasksCount,
                    total: tasks.count,
                    percentage: progressPercentage
                )
                .padding(.horizontal)
                .padding(.bottom)

                // Tab Bar
                Picker("Filter", selection: $selectedTab) {
                    Text("All (\(tasks.count))").tag(0)
                    Text("Active (\(activeTasksCount))").tag(1)
                    Text("Done (\(completedTasksCount))").tag(2)
                }
                .pickerStyle(SegmentedPickerStyle())
                .padding(.horizontal)
                .padding(.bottom, 8)

                if filteredTasks.isEmpty {
                    // Empty State
                    VStack(spacing: 16) {
                        Spacer()

                        Image(systemName: selectedTab == 2 ? "checkmark.circle" : "list.clipboard")
                            .font(.system(size: 64))
                            .foregroundColor(.gray.opacity(0.5))

                        Text(selectedTab == 2 ? "No completed tasks yet" : "No tasks yet")
                            .font(.title2)
                            .foregroundColor(.secondary)

                        if selectedTab != 2 {
                            Button("Add Task") {
                                showingAddTask = true
                            }
                            .buttonStyle(.bordered)
                        }

                        Spacer()
                    }
                } else {
                    // Tasks List
                    List {
                        ForEach(filteredTasks) { task in
                            NavigationLink(destination: TaskDetailScreen(task: task)) {
                                TaskRowView(task: task) {
                                    toggleTaskCompletion(task)
                                }
                            }
                        }
                        .onDelete(perform: deleteTasks)
                    }
                    .listStyle(PlainListStyle())
                }
            }
            .navigationTitle("Tasks")
            .toolbar {
                ToolbarItemGroup(placement: .navigationBarTrailing) {
                    Button {
                        // Show sort options
                    } label: {
                        Image(systemName: "arrow.up.arrow.down")
                    }

                    Button {
                        showingAddTask = true
                    } label: {
                        Image(systemName: "plus")
                    }
                }
            }
            .alert("New Task", isPresented: $showingAddTask) {
                TextField("Task title", text: $newTaskTitle)
                Button("Cancel", role: .cancel) {
                    newTaskTitle = ""
                }
                Button("Add") {
                    addTask()
                }
            } message: {
                Text("Enter the task title")
            }
        }
    }

    private func toggleTaskCompletion(_ task: Task) {
        if let index = tasks.firstIndex(where: { $0.id == task.id }) {
            tasks[index].isCompleted.toggle()
        }
    }

    private func deleteTasks(at offsets: IndexSet) {
        for offset in offsets {
            let task = filteredTasks[offset]
            tasks.removeAll { $0.id == task.id }
        }
    }

    private func addTask() {
        let newTask = Task(
            id: UUID().uuidString,
            title: newTaskTitle,
            description: "",
            isCompleted: false,
            priority: .medium,
            dueDate: Calendar.current.date(byAdding: .day, value: 1, to: Date()),
            category: .personal
        )
        tasks.append(newTask)
        newTaskTitle = ""
    }
}

struct TaskRowView: View {
    let task: Task
    let onToggle: () -> Void

    var body: some View {
        HStack(alignment: .top, spacing: 12) {
            // Checkbox
            Button(action: onToggle) {
                ZStack {
                    Circle()
                        .stroke(task.isCompleted ? Color.green : task.priority.color, lineWidth: 2)
                        .frame(width: 24, height: 24)

                    if task.isCompleted {
                        Circle()
                            .fill(Color.green)
                            .frame(width: 24, height: 24)

                        Image(systemName: "checkmark")
                            .font(.system(size: 12, weight: .bold))
                            .foregroundColor(.white)
                    }
                }
            }
            .buttonStyle(PlainButtonStyle())

            // Task Content
            VStack(alignment: .leading, spacing: 8) {
                // Title and Priority
                HStack {
                    Text(task.title)
                        .font(.system(size: 16, weight: .semibold))
                        .foregroundColor(task.isCompleted ? .secondary : .primary)
                        .strikethrough(task.isCompleted)

                    Spacer()

                    // Priority Badge
                    Text(task.priority.rawValue.uppercased())
                        .font(.caption2)
                        .fontWeight(.bold)
                        .padding(.horizontal, 6)
                        .padding(.vertical, 2)
                        .background(task.priority.color.opacity(0.1))
                        .foregroundColor(task.priority.color)
                        .clipShape(RoundedRectangle(cornerRadius: 6))
                }

                // Description
                if !task.description.isEmpty {
                    Text(task.description)
                        .font(.caption)
                        .foregroundColor(.secondary)
                        .strikethrough(task.isCompleted)
                        .lineLimit(2)
                }

                // Meta info
                HStack {
                    // Category
                    Text(task.category.rawValue.capitalized)
                        .font(.caption2)
                        .fontWeight(.medium)
                        .padding(.horizontal, 6)
                        .padding(.vertical, 2)
                        .background(task.category.color.opacity(0.1))
                        .foregroundColor(task.category.color)
                        .clipShape(RoundedRectangle(cornerRadius: 4))

                    if let dueDate = task.dueDate {
                        HStack(spacing: 2) {
                            Image(systemName: "clock")
                                .font(.caption2)

                            Text(task.formattedDueDate)
                                .font(.caption2)
                        }
                        .foregroundColor(task.isOverdue ? .red : .secondary)
                        .fontWeight(task.isOverdue ? .semibold : .regular)
                    }

                    Spacer()
                }
            }

            // More actions
            Menu {
                Button("Edit") {
                    // Edit task
                }

                Button("Duplicate") {
                    // Duplicate task
                }

                Button("Delete", role: .destructive) {
                    // Delete task
                }
            } label: {
                Image(systemName: "ellipsis")
                    .foregroundColor(.secondary)
            }
        }
        .padding(.vertical, 4)
    }
}

struct ProgressSummaryView: View {
    let completed: Int
    let total: Int
    let percentage: Double

    var body: some View {
        VStack(spacing: 12) {
            HStack {
                Text("Today's Progress")
                    .font(.headline)
                    .fontWeight(.bold)
                    .foregroundColor(.white)

                Spacer()

                Text("\(Int(percentage * 100))%")
                    .font(.title2)
                    .fontWeight(.bold)
                    .foregroundColor(.white)
                    .padding(.horizontal, 12)
                    .padding(.vertical, 6)
                    .background(Color.white.opacity(0.2))
                    .clipShape(Circle())
            }

            HStack {
                VStack(alignment: .leading, spacing: 4) {
                    Text("\(completed) of \(total) tasks")
                        .font(.subheadline)
                        .foregroundColor(.white.opacity(0.9))

                    ProgressView(value: percentage)
                        .progressViewStyle(LinearProgressViewStyle(tint: .white))
                        .scaleEffect(x: 1, y: 2, anchor: .center)
                }

                Spacer()
            }
        }
        .padding(20)
        .background(
            LinearGradient(
                gradient: Gradient(colors: [.blue, .blue.opacity(0.8)]),
                startPoint: .topLeading,
                endPoint: .bottomTrailing
            )
        )
        .clipShape(RoundedRectangle(cornerRadius: 16))
    }
}

enum TaskPriority: String, CaseIterable {
    case high = "high"
    case medium = "medium"
    case low = "low"

    var color: Color {
        switch self {
        case .high: return .red
        case .medium: return .orange
        case .low: return .blue
        }
    }
}

enum TaskCategory: String, CaseIterable {
    case work = "work"
    case personal = "personal"
    case health = "health"
    case learning = "learning"
    case shopping = "shopping"

    var color: Color {
        switch self {
        case .work: return .blue
        case .personal: return .green
        case .health: return .red
        case .learning: return .purple
        case .shopping: return .orange
        }
    }
}

struct Task: Identifiable {
    let id: String
    let title: String
    let description: String
    var isCompleted: Bool
    let priority: TaskPriority
    let dueDate: Date?
    let category: TaskCategory

    var isOverdue: Bool {
        guard let dueDate = dueDate, !isCompleted else { return false }
        return dueDate < Date()
    }

    var formattedDueDate: String {
        guard let dueDate = dueDate else { return "" }

        let calendar = Calendar.current
        let daysFromNow = calendar.dateComponents([.day], from: Date(), to: dueDate).day ?? 0

        if daysFromNow == 0 {
            return "Today"
        } else if daysFromNow == 1 {
            return "Tomorrow"
        } else if daysFromNow == -1 {
            return "Yesterday"
        } else if daysFromNow > 0 {
            return "In \(daysFromNow) days"
        } else {
            return "\(-daysFromNow) days ago"
        }
    }
}

#Preview {
    TasksListScreen()
}

File: snippets/screens/flutter/tasks_list_screen/1/tasks_list_screen.dart
import 'package:flutter/material.dart';

class TasksListScreen extends StatefulWidget {
  const TasksListScreen({super.key});

  @override
  State<TasksListScreen> createState() => _TasksListScreenState();
}

class _TasksListScreenState extends State<TasksListScreen>
    with SingleTickerProviderStateMixin {
  late TabController _tabController;
  final TextEditingController _taskController = TextEditingController();

  final List<Task> _tasks = [
    Task(
      id: '1',
      title: 'Complete project proposal',
      description: 'Write and submit the Q4 project proposal for the mobile app redesign',
      isCompleted: false,
      priority: TaskPriority.high,
      dueDate: DateTime.now().add(const Duration(days: 2)),
      category: TaskCategory.work,
    ),
    Task(
      id: '2',
      title: 'Buy groceries',
      description: 'Milk, bread, eggs, fruits, and vegetables for the week',
      isCompleted: false,
      priority: TaskPriority.medium,
      dueDate: DateTime.now().add(const Duration(days: 1)),
      category: TaskCategory.personal,
    ),
    Task(
      id: '3',
      title: 'Review Flutter documentation',
      description: 'Read through the latest Flutter 3.x documentation and new features',
      isCompleted: true,
      priority: TaskPriority.medium,
      dueDate: DateTime.now().subtract(const Duration(days: 1)),
      category: TaskCategory.learning,
    ),
    Task(
      id: '4',
      title: 'Plan weekend trip',
      description: 'Research destinations, book accommodation, and plan activities',
      isCompleted: false,
      priority: TaskPriority.low,
      dueDate: DateTime.now().add(const Duration(days: 7)),
      category: TaskCategory.personal,
    ),
    Task(
      id: '5',
      title: 'Call dentist appointment',
      description: 'Schedule routine checkup and cleaning for next month',
      isCompleted: false,
      priority: TaskPriority.high,
      dueDate: DateTime.now().add(const Duration(days: 1)),
      category: TaskCategory.health,
    ),
  ];

  @override
  void initState() {
    super.initState();
    _tabController = TabController(length: 3, vsync: this);
  }

  List<Task> get _filteredTasks {
    switch (_tabController.index) {
      case 0: // All
        return _tasks;
      case 1: // Active
        return _tasks.where((task) => !task.isCompleted).toList();
      case 2: // Completed
        return _tasks.where((task) => task.isCompleted).toList();
      default:
        return _tasks;
    }
  }

  @override
  Widget build(BuildContext context) {
    final activeTasks = _tasks.where((task) => !task.isCompleted).length;
    final completedTasks = _tasks.where((task) => task.isCompleted).length;

    return Scaffold(
      appBar: AppBar(
        title: const Text('Tasks'),
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        actions: [
          IconButton(
            icon: const Icon(Icons.sort),
            onPressed: () => _showSortDialog(),
          ),
        ],
        bottom: TabBar(
          controller: _tabController,
          onTap: (index) => setState(() {}),
          tabs: [
            Tab(text: 'All (${_tasks.length})'),
            Tab(text: 'Active ($activeTasks)'),
            Tab(text: 'Done ($completedTasks)'),
          ],
        ),
      ),
      body: Column(
        children: [
          // Progress Summary
          Container(
            width: double.infinity,
            margin: const EdgeInsets.all(16),
            padding: const EdgeInsets.all(20),
            decoration: BoxDecoration(
              gradient: LinearGradient(
                colors: [Colors.blue.shade400, Colors.blue.shade600],
                begin: Alignment.topLeft,
                end: Alignment.bottomRight,
              ),
              borderRadius: BorderRadius.circular(16),
            ),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                const Text(
                  'Today\'s Progress',
                  style: TextStyle(
                    color: Colors.white,
                    fontSize: 18,
                    fontWeight: FontWeight.bold,
                  ),
                ),
                const SizedBox(height: 12),
                Row(
                  children: [
                    Expanded(
                      child: Column(
                        crossAxisAlignment: CrossAxisAlignment.start,
                        children: [
                          Text(
                            '$completedTasks of ${_tasks.length} tasks',
                            style: const TextStyle(
                              color: Colors.white70,
                              fontSize: 14,
                            ),
                          ),
                          const SizedBox(height: 8),
                          LinearProgressIndicator(
                            value: _tasks.isEmpty ? 0 : completedTasks / _tasks.length,
                            backgroundColor: Colors.white30,
                            valueColor: const AlwaysStoppedAnimation<Color>(Colors.white),
                          ),
                        ],
                      ),
                    ),
                    const SizedBox(width: 16),
                    CircleAvatar(
                      radius: 20,
                      backgroundColor: Colors.white,
                      child: Text(
                        '${(_tasks.isEmpty ? 0 : (completedTasks / _tasks.length * 100)).round()}%',
                        style: TextStyle(
                          color: Colors.blue.shade600,
                          fontWeight: FontWeight.bold,
                          fontSize: 12,
                        ),
                      ),
                    ),
                  ],
                ),
              ],
            ),
          ),

          // Task List
          Expanded(
            child: TabBarView(
              controller: _tabController,
              children: [
                _buildTasksList(),
                _buildTasksList(),
                _buildTasksList(),
              ],
            ),
          ),
        ],
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () => _showAddTaskDialog(),
        child: const Icon(Icons.add),
      ),
    );
  }

  Widget _buildTasksList() {
    final tasks = _filteredTasks;

    if (tasks.isEmpty) {
      return Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Icon(
              _tabController.index == 2 ? Icons.task_alt : Icons.assignment,
              size: 64,
              color: Colors.grey[400],
            ),
            const SizedBox(height: 16),
            Text(
              _tabController.index == 2
                  ? 'No completed tasks yet'
                  : 'No tasks yet',
              style: Theme.of(context).textTheme.titleMedium?.copyWith(
                    color: Colors.grey[600],
                  ),
            ),
            if (_tabController.index != 2) ...[
              const SizedBox(height: 8),
              TextButton.icon(
                onPressed: () => _showAddTaskDialog(),
                icon: const Icon(Icons.add),
                label: const Text('Add Task'),
              ),
            ],
          ],
        ),
      );
    }

    return ListView.builder(
      padding: const EdgeInsets.symmetric(horizontal: 16),
      itemCount: tasks.length,
      itemBuilder: (context, index) {
        final task = tasks[index];
        final isOverdue = !task.isCompleted &&
                         task.dueDate != null &&
                         task.dueDate!.isBefore(DateTime.now());

        return Card(
          margin: const EdgeInsets.only(bottom: 8),
          child: InkWell(
            onTap: () => _navigateToTaskDetail(task),
            borderRadius: BorderRadius.circular(12),
            child: Padding(
              padding: const EdgeInsets.all(16),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Row(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      // Checkbox
                      InkWell(
                        onTap: () => _toggleTaskCompletion(task),
                        borderRadius: BorderRadius.circular(20),
                        child: Container(
                          width: 24,
                          height: 24,
                          decoration: BoxDecoration(
                            shape: BoxShape.circle,
                            border: Border.all(
                              color: task.isCompleted
                                  ? Colors.green
                                  : _getPriorityColor(task.priority),
                              width: 2,
                            ),
                            color: task.isCompleted ? Colors.green : Colors.transparent,
                          ),
                          child: task.isCompleted
                              ? const Icon(
                                  Icons.check,
                                  size: 16,
                                  color: Colors.white,
                                )
                              : null,
                        ),
                      ),
                      const SizedBox(width: 12),

                      // Task Content
                      Expanded(
                        child: Column(
                          crossAxisAlignment: CrossAxisAlignment.start,
                          children: [
                            Row(
                              children: [
                                Expanded(
                                  child: Text(
                                    task.title,
                                    style: TextStyle(
                                      fontWeight: FontWeight.w600,
                                      fontSize: 16,
                                      decoration: task.isCompleted
                                          ? TextDecoration.lineThrough
                                          : null,
                                      color: task.isCompleted
                                          ? Colors.grey
                                          : null,
                                    ),
                                  ),
                                ),
                                // Priority Badge
                                Container(
                                  padding: const EdgeInsets.symmetric(
                                    horizontal: 8,
                                    vertical: 2,
                                  ),
                                  decoration: BoxDecoration(
                                    color: _getPriorityColor(task.priority).withOpacity(0.1),
                                    borderRadius: BorderRadius.circular(12),
                                  ),
                                  child: Text(
                                    _getPriorityName(task.priority),
                                    style: TextStyle(
                                      fontSize: 10,
                                      fontWeight: FontWeight.w600,
                                      color: _getPriorityColor(task.priority),
                                    ),
                                  ),
                                ),
                              ],
                            ),

                            if (task.description.isNotEmpty) ...[
                              const SizedBox(height: 4),
                              Text(
                                task.description,
                                style: TextStyle(
                                  color: Colors.grey[600],
                                  fontSize: 14,
                                  decoration: task.isCompleted
                                      ? TextDecoration.lineThrough
                                      : null,
                                ),
                                maxLines: 2,
                                overflow: TextOverflow.ellipsis,
                              ),
                            ],

                            const SizedBox(height: 8),

                            // Task Meta
                            Row(
                              children: [
                                // Category
                                Container(
                                  padding: const EdgeInsets.symmetric(
                                    horizontal: 8,
                                    vertical: 2,
                                  ),
                                  decoration: BoxDecoration(
                                    color: _getCategoryColor(task.category).withOpacity(0.1),
                                    borderRadius: BorderRadius.circular(8),
                                  ),
                                  child: Text(
                                    _getCategoryName(task.category),
                                    style: TextStyle(
                                      fontSize: 10,
                                      fontWeight: FontWeight.w500,
                                      color: _getCategoryColor(task.category),
                                    ),
                                  ),
                                ),

                                if (task.dueDate != null) ...[
                                  const SizedBox(width: 8),
                                  Icon(
                                    Icons.schedule,
                                    size: 14,
                                    color: isOverdue ? Colors.red : Colors.grey[600],
                                  ),
                                  const SizedBox(width: 4),
                                  Text(
                                    _formatDueDate(task.dueDate!),
                                    style: TextStyle(
                                      fontSize: 12,
                                      color: isOverdue ? Colors.red : Colors.grey[600],
                                      fontWeight: isOverdue ? FontWeight.w600 : null,
                                    ),
                                  ),
                                ],

                                const Spacer(),

                                // More Options
                                PopupMenuButton<String>(
                                  onSelected: (value) => _handleTaskAction(value, task),
                                  itemBuilder: (context) => [
                                    const PopupMenuItem(
                                      value: 'edit',
                                      child: ListTile(
                                        leading: Icon(Icons.edit, size: 20),
                                        title: Text('Edit'),
                                        dense: true,
                                      ),
                                    ),
                                    const PopupMenuItem(
                                      value: 'duplicate',
                                      child: ListTile(
                                        leading: Icon(Icons.copy, size: 20),
                                        title: Text('Duplicate'),
                                        dense: true,
                                      ),
                                    ),
                                    const PopupMenuItem(
                                      value: 'delete',
                                      child: ListTile(
                                        leading: Icon(Icons.delete, size: 20, color: Colors.red),
                                        title: Text('Delete', style: TextStyle(color: Colors.red)),
                                        dense: true,
                                      ),
                                    ),
                                  ],
                                  child: Icon(
                                    Icons.more_vert,
                                    color: Colors.grey[600],
                                    size: 20,
                                  ),
                                ),
                              ],
                            ),
                          ],
                        ),
                      ),
                    ],
                  ),
                ],
              ),
            ),
          ),
        );
      },
    );
  }

  Color _getPriorityColor(TaskPriority priority) {
    switch (priority) {
      case TaskPriority.high:
        return Colors.red;
      case TaskPriority.medium:
        return Colors.orange;
      case TaskPriority.low:
        return Colors.blue;
    }
  }

  String _getPriorityName(TaskPriority priority) {
    switch (priority) {
      case TaskPriority.high:
        return 'HIGH';
      case TaskPriority.medium:
        return 'MED';
      case TaskPriority.low:
        return 'LOW';
    }
  }

  Color _getCategoryColor(TaskCategory category) {
    switch (category) {
      case TaskCategory.work:
        return Colors.blue;
      case TaskCategory.personal:
        return Colors.green;
      case TaskCategory.health:
        return Colors.red;
      case TaskCategory.learning:
        return Colors.purple;
      case TaskCategory.shopping:
        return Colors.orange;
    }
  }

  String _getCategoryName(TaskCategory category) {
    switch (category) {
      case TaskCategory.work:
        return 'Work';
      case TaskCategory.personal:
        return 'Personal';
      case TaskCategory.health:
        return 'Health';
      case TaskCategory.learning:
        return 'Learning';
      case TaskCategory.shopping:
        return 'Shopping';
    }
  }

  String _formatDueDate(DateTime date) {
    final now = DateTime.now();
    final difference = date.difference(now).inDays;

    if (difference == 0) {
      return 'Today';
    } else if (difference == 1) {
      return 'Tomorrow';
    } else if (difference == -1) {
      return 'Yesterday';
    } else if (difference > 0) {
      return 'In $difference days';
    } else {
      return '${-difference} days ago';
    }
  }

  void _toggleTaskCompletion(Task task) {
    setState(() {
      task.isCompleted = !task.isCompleted;
    });

    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(
        content: Text(
          task.isCompleted
              ? 'Task completed!'
              : 'Task marked as active',
        ),
        duration: const Duration(seconds: 2),
      ),
    );
  }

  void _navigateToTaskDetail(Task task) {
    Navigator.of(context).push(
      MaterialPageRoute(
        builder: (context) => TaskDetailScreen(task: task),
      ),
    );
  }

  void _handleTaskAction(String action, Task task) {
    switch (action) {
      case 'edit':
        _editTask(task);
        break;
      case 'duplicate':
        _duplicateTask(task);
        break;
      case 'delete':
        _deleteTask(task);
        break;
    }
  }

  void _editTask(Task task) {
    ScaffoldMessenger.of(context).showSnackBar(
      const SnackBar(content: Text('Edit task functionality would be implemented here')),
    );
  }

  void _duplicateTask(Task task) {
    setState(() {
      _tasks.add(Task(
        id: '${_tasks.length + 1}',
        title: '${task.title} (Copy)',
        description: task.description,
        isCompleted: false,
        priority: task.priority,
        dueDate: task.dueDate,
        category: task.category,
      ));
    });

    ScaffoldMessenger.of(context).showSnackBar(
      const SnackBar(content: Text('Task duplicated!')),
    );
  }

  void _deleteTask(Task task) {
    showDialog(
      context: context,
      builder: (context) => AlertDialog(
        title: const Text('Delete Task'),
        content: Text('Are you sure you want to delete "${task.title}"?'),
        actions: [
          TextButton(
            onPressed: () => Navigator.of(context).pop(),
            child: const Text('Cancel'),
          ),
          TextButton(
            onPressed: () {
              setState(() {
                _tasks.removeWhere((t) => t.id == task.id);
              });
              Navigator.of(context).pop();
              ScaffoldMessenger.of(context).showSnackBar(
                const SnackBar(content: Text('Task deleted')),
              );
            },
            style: TextButton.styleFrom(foregroundColor: Colors.red),
            child: const Text('Delete'),
          ),
        ],
      ),
    );
  }

  void _showAddTaskDialog() {
    showDialog(
      context: context,
      builder: (context) => AlertDialog(
        title: const Text('Add New Task'),
        content: TextField(
          controller: _taskController,
          decoration: const InputDecoration(
            hintText: 'Enter task title...',
            border: OutlineInputBorder(),
          ),
          autofocus: true,
        ),
        actions: [
          TextButton(
            onPressed: () {
              _taskController.clear();
              Navigator.of(context).pop();
            },
            child: const Text('Cancel'),
          ),
          ElevatedButton(
            onPressed: () {
              if (_taskController.text.trim().isNotEmpty) {
                setState(() {
                  _tasks.add(Task(
                    id: '${_tasks.length + 1}',
                    title: _taskController.text.trim(),
                    description: '',
                    isCompleted: false,
                    priority: TaskPriority.medium,
                    dueDate: DateTime.now().add(const Duration(days: 1)),
                    category: TaskCategory.personal,
                  ));
                });
                _taskController.clear();
                Navigator.of(context).pop();
                ScaffoldMessenger.of(context).showSnackBar(
                  const SnackBar(content: Text('Task added!')),
                );
              }
            },
            child: const Text('Add'),
          ),
        ],
      ),
    );
  }

  void _showSortDialog() {
    ScaffoldMessenger.of(context).showSnackBar(
      const SnackBar(content: Text('Sort options would be implemented here')),
    );
  }

  @override
  void dispose() {
    _tabController.dispose();
    _taskController.dispose();
    super.dispose();
  }
}

enum TaskPriority { high, medium, low }
enum TaskCategory { work, personal, health, learning, shopping }

class Task {
  final String id;
  final String title;
  final String description;
  bool isCompleted;
  final TaskPriority priority;
  final DateTime? dueDate;
  final TaskCategory category;

  Task({
    required this.id,
    required this.title,
    required this.description,
    required this.isCompleted,
    required this.priority,
    this.dueDate,
    required this.category,
  });
}

// This would typically be in a separate file
class TaskDetailScreen extends StatelessWidget {
  final Task task;

  const TaskDetailScreen({super.key, required this.task});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(task.title),
      ),
      body: const Center(
        child: Text('Task Detail Screen'),
      ),
    );
  }
}